File: backend/app/services/dashboard_service.py
from sqlalchemy.orm import Session

from app.models.book import Book
from app.models.note import Note
from app.models.quiz import Quiz
from app.models.current_affair import CurrentAffair


class DashboardService:

    # =====================================================
    # Dashboard Summary
    # =====================================================

    @staticmethod
    def get_dashboard_summary(db: Session):
        return {
            "books": db.query(Book).count(),
            "notes": db.query(Note).count(),
            "quizzes": db.query(Quiz).count(),
            "current_affairs": db.query(CurrentAffair).count(),
        }

    # =====================================================
    # Subject Statistics
    # =====================================================

    @staticmethod
    def subject_statistics(db: Session):
        return DashboardService.get_dashboard_summary(db)

    # =====================================================
    # Featured Content
    # =====================================================

    @staticmethod
    def featured_content(db: Session):
        return (
            db.query(CurrentAffair)
            .filter(CurrentAffair.featured == True)
            .order_by(CurrentAffair.publish_date.desc())
            .limit(10)
            .all()
        )

    # =====================================================
    # Reading Progress (Mock)
    # =====================================================

    @staticmethod
    def reading_progress(db: Session):
        return {
            "completed_books": 12,
            "completed_notes": 34,
            "overall_progress": 72,
        }

    # =====================================================
    # Recent Books
    # =====================================================

    @staticmethod
    def recent_books(db: Session, limit: int = 5):
        return (
            db.query(Book)
            .order_by(Book.id.desc())
            .limit(limit)
            .all()
        )

    # =====================================================
    # Recent Notes
    # =====================================================

    @staticmethod
    def recent_notes(db: Session, limit: int = 5):
        return (
            db.query(Note)
            .order_by(Note.id.desc())
            .limit(limit)
            .all()
        )

    # =====================================================
    # Recent Quizzes
    # =====================================================

    @staticmethod
    def recent_quizzes(db: Session, limit: int = 5):
        return (
            db.query(Quiz)
            .order_by(Quiz.id.desc())
            .limit(limit)
            .all()
        )

    # =====================================================
    # Recent Current Affairs
    # =====================================================

    @staticmethod
    def recent_current_affairs(db: Session, limit: int = 5):
        return (
            db.query(CurrentAffair)
            .order_by(CurrentAffair.publish_date.desc())
            .limit(limit)
            .all()
        )

    # =====================================================
    # Continue Reading (Mock)
    # =====================================================

    @staticmethod
    def continue_reading(db: Session, user_id: int):
        return {
            "book": "Indian Polity",
            "chapter": "Fundamental Rights",
            "progress": 72,
        }

    # =====================================================
    # Weekly Progress (Mock)
    # =====================================================

    @staticmethod
    def weekly_progress(db: Session):
        return [
            {"day": "Mon", "hours": 2},
            {"day": "Tue", "hours": 3},
            {"day": "Wed", "hours": 2.5},
            {"day": "Thu", "hours": 4},
            {"day": "Fri", "hours": 3},
            {"day": "Sat", "hours": 5},
            {"day": "Sun", "hours": 2},
        ]

    # =====================================================
    # Recent Quiz (Mock)
    # =====================================================

    @staticmethod
    def recent_quiz(db: Session, user_id: int):
        return {
            "quiz_id": 101,
            "title": "General Studies Mock Test",
            "score": 86,
            "correct_answers": 43,
            "total_questions": 50,
        }

File: frontend/src/services/dashboardService.ts
import api from "@/lib/api";

export interface DashboardSummary {
    books: number;
    notes: number;
    quizzes: number;
    current_affairs: number;
}

export interface ReadingProgress {
    completed_books: number;
    completed_notes: number;
    overall_progress: number;
}

export interface DashboardStatistics {
    books: number;
    notes: number;
    quizzes: number;
    current_affairs: number;
}

export interface WeeklyProgress {
    day: string;
    hours: number;
}

export interface ContinueReading {
    book: string;
    chapter: string;
    progress: number;
}

export interface RecentQuiz {
    quiz_id: number;
    title: string;
    score: number;
    correct_answers: number;
    total_questions: number;
}

export const DashboardService = {

    // =====================================================
    // Dashboard Summary
    // =====================================================

    async summary() {
        return api.get<DashboardSummary>(
            "/dashboard/summary"
        );
    },

    // =====================================================
    // Statistics
    // =====================================================

    async statistics() {
        return api.get<DashboardStatistics>(
            "/dashboard/statistics"
        );
    },

    // =====================================================
    // Reading Progress
    // =====================================================

    async progress() {
        return api.get<ReadingProgress>(
            "/dashboard/progress"
        );
    },

    // =====================================================
    // Featured Content
    // =====================================================

    async featured() {
        return api.get(
            "/dashboard/featured"
        );
    },

    // =====================================================
    // Recent Books
    // =====================================================

    async recentBooks(limit = 5) {
        return api.get(
            "/dashboard/recent/books",
            {
                params: { limit },
            }
        );
    },

    // =====================================================
    // Recent Notes
    // =====================================================

    async recentNotes(limit = 5) {
        return api.get(
            "/dashboard/recent/notes",
            {
                params: { limit },
            }
        );
    },

    // =====================================================
    // Recent Quizzes
    // =====================================================

    async recentQuizzes(limit = 5) {
        return api.get(
            "/dashboard/recent/quizzes",
            {
                params: { limit },
            }
        );
    },

    // =====================================================
    // Recent Current Affairs
    // =====================================================

    async recentCurrentAffairs(limit = 5) {
        return api.get(
            "/dashboard/recent/current-affairs",
            {
                params: { limit },
            }
        );
    },

    // =====================================================
    // Continue Reading
    // =====================================================

    async continueReading() {
        return api.get<ContinueReading>(
            "/dashboard/continue-reading"
        );
    },

    // =====================================================
    // Weekly Progress
    // =====================================================

    async weeklyProgress() {
        return api.get<WeeklyProgress[]>(
            "/dashboard/weekly-progress"
        );
    },

    // =====================================================
    // Recent Quiz
    // =====================================================

    async recentQuiz() {
        return api.get<RecentQuiz>(
            "/dashboard/recent-quiz"
        );
    },

    // =====================================================
    // Refresh Dashboard
    // =====================================================

    async refresh() {

        const [
            summary,
            statistics,
            progress,
            featured,
            books,
            notes,
            quizzes,
            currentAffairs,
            continueReading,
            weeklyProgress,
            recentQuiz,
        ] = await Promise.all([
            this.summary(),
            this.statistics(),
            this.progress(),
            this.featured(),
            this.recentBooks(),
            this.recentNotes(),
            this.recentQuizzes(),
            this.recentCurrentAffairs(),
            this.continueReading(),
            this.weeklyProgress(),
            this.recentQuiz(),
        ]);

        return {
            summary: summary.data,
            statistics: statistics.data,
            progress: progress.data,
            featured: featured.data,
            books: books.data,
            notes: notes.data,
            quizzes: quizzes.data,
            currentAffairs: currentAffairs.data,
            continueReading: continueReading.data,
            weeklyProgress: weeklyProgress.data,
            recentQuiz: recentQuiz.data,
        };

    },

};

export default DashboardService;